### DISCUSSION TITLE
How to return the response from an asynchronous call

### DISCUSSION THREAD

**Discussion by toshikijp** [andatoshiki/toshiki-discuss#27] [Q&A] [answered]
I have a function `foo` which makes an asynchronous request. How can I return the response/result from `foo`?

I am trying to return the value from the callback, as well as assigning the result to a local variable inside the function and returning that one, but none of those ways actually return the response (they all return `undefined` or whatever the initial value of the variable `result` is).

**Example of an asynchronous function that accepts a callback** (using jQuery's `ajax` function)

```js
function foo() {
    var result;

    $.ajax({
        url: '...',
        success: function(response) {
            result = response;
            // return response; // <- I tried that one as well
        }
    });

    return result; // It always returns `undefined`
}
```

**Example using Node.js:**

```js
function foo() {
    var result;

    fs.readFile("path/to/file", function(err, data) {
        result = data;
        // return data; // <- I tried that one as well
    });

    return result; // It always returns `undefined`
}
```

**Example using the `then` block of a promise:**

```js
function foo() {
    var result;

    fetch(url).then(function(response) {
        result = response;
        // return response; // <- I tried that one as well
    });

    return result; // It always returns `undefined`
}
```

**Comment by andatoshiki** [comment, marked as answer]
# If you're *not* using jQuery in your code, this answer is for you

Your code should be something along the lines of this:

```js
function foo() {
    var httpRequest = new XMLHttpRequest();
    httpRequest.open('GET', "/echo/json");
    httpRequest.send();
    return httpRequest.responseText;
}

var result = foo(); // Always ends up being 'undefined'
```

[[Felix Kling did a fine job](https://stackoverflow.com/questions/14220321/how-do-i-return-the-response-from-an-asynchronous-call/14220323#14220323)](https://stackoverflow.com/questions/14220321/how-do-i-return-the-response-from-an-asynchronous-call/14220323#14220323) writing an answer for people using jQuery for AJAX, but I've decided to provide an alternative for people who aren't.

([Note, for those using the new `fetch` API, Angular or promises I've added another answer below](https://stackoverflow.com/a/30180679/1348195))

------

# What you're facing

This is a short summary of "Explanation of the problem" from the other answer, if you're not sure after reading this, read that.

The **A** in AJAX stands for **asynchronous**. That means sending the request (or rather receiving the response) is taken out of the normal execution flow. In your example, [`[.send](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest#send())`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest#send()) returns immediately and the next statement, `return result;`, is executed before the function you passed as `success` callback was even called.

This means when you're returning, the listener you've defined did not execute yet, which means the value you're returning has not been defined.

Here is a simple analogy:

```js
function getFive(){
    var a;
    setTimeout(function(){
         a=5;
    },10);
    return a;
}
```

[[(Fiddle)](http://jsfiddle.net/7RK3k/)](http://jsfiddle.net/7RK3k/)

The value of `a` returned is `undefined` since the `a=5` part has not executed yet. AJAX acts like this, you're returning the value before the server got the chance to tell your browser what that value is.

One possible solution to this problem is to code *re-actively* , telling your program what to do when the calculation completed.

```js
function onComplete(a){ // When the code completes, do this
    alert(a);
}

function getFive(whenDone){
    var a;
    setTimeout(function(){
         a=5;
         whenDone(a);
    },10);
}
```

This is called [[CPS](http://en.wikipedia.org/wiki/Continuation-passing_style)](http://en.wikipedia.org/wiki/Continuation-passing_style). Basically, we're passing `getFive` an action to perform when it completes, we're telling our code how to react when an event completes (like our AJAX call, or in this case the timeout).

Usage would be:

```js
getFive(onComplete);
```

Which should alert "5" to the screen. [[(Fiddle)](http://jsfiddle.net/PAjZR/)](http://jsfiddle.net/PAjZR/).

# Possible solutions

There are basically two ways how to solve this:

1. Make the AJAX call synchronous (let’s call it SJAX).
2. Restructure your code to work properly with callbacks.

## 1. Synchronous AJAX - Don't do it!!

As for synchronous AJAX, **don't do it!** Felix's answer raises some compelling arguments about why it's a bad idea. To sum it up, it'll freeze the user's browser until the server returns the response and create a very bad user experience. Here is another short summary taken from MDN on why:

> XMLHttpRequest supports both synchronous and asynchronous communications. In general, however, asynchronous requests should be preferred to synchronous requests for performance reasons.
>
> In short, synchronous requests block the execution of code... ...this can cause serious issues...

If you *have* to do it, you can pass a flag. [[Here is how](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests#Synchronous_request)](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests#Synchronous_request):

```js
var request = new XMLHttpRequest();
request.open('GET', 'yourURL', false);  // `false` makes the request synchronous
request.send(null);

if (request.status === 200) {// That's HTTP for 'ok'
  console.log(request.responseText);
}
```

## 2. Restructure code

Let your function accept a callback. In the example code `foo` can be made to accept a callback. We'll be telling our code how to *react* when `foo` completes.

So:

```js
var result = foo();
// Code that depends on `result` goes here
```

Becomes:

```js
foo(function(result) {
    // Code that depends on `result`
});
```

Here we passed an anonymous function, but we could just as easily pass a reference to an existing function, making it look like:

```js
function myHandler(result) {
    // Code that depends on `result`
}
foo(myHandler);
```

For more details on how this sort of callback design is done, check Felix's answer.

Now, let's define foo itself to act accordingly

```js
function foo(callback) {
    var httpRequest = new XMLHttpRequest();
    httpRequest.onload = function(){ // When the request is loaded
       callback(httpRequest.responseText);// We're calling our method
    };
    httpRequest.open('GET', "/echo/json");
    httpRequest.send();
}
```

[[(fiddle)](http://jsfiddle.net/DAcWT/)](http://jsfiddle.net/DAcWT/)

We have now made our *foo* function accept an action to run when the AJAX completes successfully. We can extend this further by checking if the response status is not 200 and acting accordingly (create a fail handler and such). Effectively it is solving our issue.

If you're still having a hard time understanding this, [[read the AJAX getting started guide](https://developer.mozilla.org/en-US/docs/AJAX/Getting_Started)](https://developer.mozilla.org/en-US/docs/AJAX/Getting_Started) at MDN.
